Add tests

Dragorn421 opened the pull request:
Start adding tests

The current "test" stuff:
```
$ make -C test
make: Entering directory '/home/dragorn421/Documents/pygfxd/test'
./test.py
.........
----------------------------------------------------------------------
Ran 9 tests in 0.011s

OK
make: Leaving directory '/home/dragorn421/Documents/pygfxd/test'
```

Dragorn421 committed:
Update python bindings to libgfxd 7dd7fa1

libgfxd 7dd7fa16227f6d4c336abfda353d01f833334413

Update/add `gfxd_light_callback`, `gfxd_lightsn_callback`
Add binding for `gfxd_foreach_pkt`
Update description for `gfxd_macro_fn`

Dragorn421 committed:
Fix https://github.com/Thar0/pygfxd/issues/5

Dragorn421 committed:
Some test setup

Thar0 commented on test/gbi.h:
```diff
@@ -0,0 +1 @@
+#include "../libgfxd/gbi.h"
```
Could an include path for libgfxd be specified in the makefile rather than this dummy file?

Dragorn421 replied:
The idea with this wrapper .h would also be to let someone try another gbi.h, since libgfxd's is glank's i.e. not sgi's / decomp's

But yeah this can be changed idk

Thar0 commented on test/Makefile:
```diff
@@ -0,0 +1,20 @@
+# Compile C files and dump their .data section to binary
+
+CC := mips-linux-gnu-gcc
+OBJCOPY := mips-linux-gnu-objcopy
+NM := mips-linux-gnu-nm
```
```suggestion
CROSS := mips-linux-gnu-

CC := $(CROSS)gcc
OBJCOPY := $(CROSS)objcopy
NM := $(CROSS)nm
```
This suggestion was made by `mips64-ultra-elf-` gang

Thar0 commented on test/test.py:
```diff
@@ -0,0 +1,120 @@
+from pathlib import Path
```
```suggestion
#!/usr/bin/env python3
from pathlib import Path
```

Thar0 reviewed (commented):
Rough thoughts on the current stuff

I like the idea of hosting simple tests from our own C files. Depending on how many tests there ends up being it would be nice to split each test into it's own python file, but this is fine for now. `test.py` in root can probably also go.

Three small nits

Dragorn421 committed:
Fix/update `gfx_ucode` structure in py binding

Dragorn421 committed:
remove test.py, shebang gang

Dragorn421 committed:
Introduce unittest usage

Dragorn421 committed:
Add CROSS as binutils prefix

Dragorn421 committed:
git subrepo pull (merge) libgfxd

subrepo:
  subdir:   "libgfxd"
  merged:   "7dd7fa1"
upstream:
  origin:   "https://github.com/glankk/libgfxd"
  branch:   "master"
  commit:   "7dd7fa1"
git-subrepo:
  version:  "0.4.5"
  origin:   "git@github.com:ingydotnet/git-subrepo.git"
  commit:   "dbb99be"

Dragorn421 committed:
Add `test_gfxd_vtx_callback`

Dragorn421 commented on test/test.py:
```diff
@@ -0,0 +1,355 @@
+            with self.subTest(sym):
+                gfxd_input_buffer(data)
+
+                gfxd_target(gfxd_f3dex2)
+                gfxd_endian(GfxdEndian.big, 4)
```
not sure about putting this everywhere :shrug: 
Probably deal with it if/when someone uses something other than the Superior :tm: Ucode f3dzex

Dragorn421 commented on test/test.py:
```diff
@@ -0,0 +1,355 @@
+                    verts.append((vtx, num))
+                    return 0
+
+                try:
+                    gfxd_vtx_callback(callback)
+
+                    gfxd_execute()
+                finally:
+                    def callback(vtx, num):
+                        return 0
+                    gfxd_vtx_callback(callback)
```
So this (try: ... finally: clear callback) is kind of bad I think, tests need to clear the state or it will mess with other states since libgfxd "keeps running"/is the same instance

Not sure what to do about it. "As long as it works" though is imo fine for tests

I only spotted this issue from adding this vtx callback test case, I should probably go over other callbacks and do the same

------

Also `gfxd_vtx_callback(None)` should eventually be made to work as you'd expect (afaict there currently is no way to unset a callback)

Dragorn421 commented on test/test.py:
```diff
@@ -0,0 +1,355 @@
+            with self.subTest(sym):
+                remaining_data = [data]
+
+                def callback(bufP, count):
+                    newdata = remaining_data[0][:count]
+                    from ctypes import c_char
+
+                    buftype = c_char * len(newdata)
+                    buf = buftype.from_address(bufP)
+                    buf[: len(newdata)] = newdata
+                    remaining_data[0] = remaining_data[0][len(newdata) :]
+                    return len(newdata)
```
note: this could be improved on the pygfxd wrapper side

Dragorn421 committed:
fix unsetting buf/cb (wip), rework anti-gc buf/cb dict, fix https://github.com/Thar0/pygfxd/issues/10

Dragorn421 committed:
Merge branch 'master' into dev

Dragorn421 committed:
Fix somehow removed actual call to output_buffer

Dragorn421 committed:
Merge branch 'master' into dev

Dragorn421 committed:
Fix: replace del with .pop(..., None) to fix KeyError if no callback is stored... of course...

Dragorn421 committed:
Merge branch 'master' into dev

Dragorn421 committed:
Fix `gfxd_arg_value`, `gfxd_value_by_type` bindings

Dragorn421 committed:
Merge branch 'master' into dev

Dragorn421 committed:
more tests

Dragorn421 committed:
fix

Dragorn421 committed:
Have gfxd_vtx_callback binding take None

Dragorn421 committed:
Merge branch 'master' into dev

Dragorn421 committed:
test nulling "missing" callback

Dragorn421 committed:
fix

Dragorn421 committed:
Merge branch 'master' into dev

Dragorn421 committed:
Merge remote-tracking branch 'upstream/master' into dev

Dragorn421 committed:
Merge remote-tracking branch 'upstream/master' into dev

Dragorn421 committed:
revert pygfxd.py

Dragorn421 committed:
make CI run tests, attempt 1

Dragorn421 committed:
run tests in CI, attempt 2

Dragorn421 committed:
run tests in CI, attempt 3

Dragorn421 committed:
run tests in CI, attempt 4

Dragorn421 committed:
run tests in CI, attempt 5

Dragorn421 committed:
Fix unsetting callbacks: pass None

Dragorn421 committed:
Merge branch 'fix_unsetting_callbacks' into dev

Dragorn421 committed:
fix fix

Dragorn421 committed:
Merge remote-tracking branch 'upstream/master' into dev
